File: Politio/src/scripts/Player.ts
export class Player {
    private _email: string
    private _title: string
    private _avatar: string
    private _coinsTotal: number
    private _kudosTotal: number
    private _gamesTotal: number
    private _kudosOverall: number
    private _modeChoices: number[]
    private _standingChoices: number[]
    private _unlockedTitleFirstWords: string[]
    private _unlockedTitleSecondWords: string[]
    private _unlockedAvatars: string[]
    private _strikes: number
    theme: string

    constructor (email: string, title: string, avatar: string, coinsTotal: number, kudosTotal: number,
                gamesTotal: number, kudosOverall: number, modeChoices: number[], standingChoices: number[], 
                unlockedTitleFirstWords: string[], unlockedTitleSecondWords: string[], 
                unlockedAvatars: string[], strikes: number, theme: string) {
        this._email = email
        this._title = title || ''
        this._avatar = avatar || ''
        this._coinsTotal = coinsTotal || 0
        this._kudosTotal = kudosTotal || 0
        this._gamesTotal = gamesTotal || 0
        this._kudosOverall = kudosOverall || 0
        this._modeChoices = modeChoices || [0, 0, 0]
        this._standingChoices = standingChoices || [0, 0, 0, 0]
        this._unlockedTitleFirstWords = unlockedTitleFirstWords || ['']
        this._unlockedTitleSecondWords = unlockedTitleSecondWords || ['']
        this._unlockedAvatars = unlockedAvatars || ['']
        this._strikes = strikes || 0
        this.theme = theme || 'light'
    }

    /*
        standingChoices[0] = number of times an authoritarian standing was chosen
        standingChoices[1] = number of times a left standing was chosen
        standingChoices[2] = number of times a libertarian standing was chosen
        standingChoices[3] = number of times a right standing was chosen
    */
    get standingActual() {

        let x = this._standingChoices[1] / this._standingChoices[3];
        let y = this._standingChoices[0] / this._standingChoices[2];
        let standing = "";

        if (x > 1.3)
            standing += "Left ";
        else if (x < 0.7)
            standing += "Right ";
        if (y > 1.3)
            standing += "Authoritarian";
        else if (y < 0.7)
            standing += "Libertarian";

        return standing;
    }

    incAuthoritarian() { this._standingChoices[0]++ }
    incLeft() { this._standingChoices[1]++ }
    incLibertarian() { this._standingChoices[2]++ }
    incRight() { this._standingChoices[3]++ }

    /*
        modeChoices[0] = the number of times the player has played 'Battle Royal'
        modeChoices[1] = the number of times the player has played 'Debate'
        modeChoices[2] = the number of times the player has played 'Jury'
    */
    get favoriteMode () {
        if (this._gamesTotal == 0)
            return "Play a game first!"
        if (this._modeChoices[0] > this._modeChoices[1] &&
            this._modeChoices[0] > this._modeChoices[2])
            return "Battle Royal"
        else if (this._modeChoices[1] > this._modeChoices[0] && 
            this._modeChoices[1] > this._modeChoices[2])
            return "Debate"
        else
            return "Jury"
    }

    addTitleFirstWords(newWords: string[]) {
        if (newWords.length != 0)
            for (let word in newWords)
                this._unlockedTitleFirstWords.push(word)
    }

    addTitleSecondWords(newWords: string[]) {
        if (newWords.length != 0)
            for (let word in newWords)
                this._unlockedTitleSecondWords.push(word)
    }

    addAvatar(newAvatar: string) {
        if (newAvatar != null && newAvatar != '')
            this._unlockedAvatars.push(newAvatar)
    }

    addCoins(coinCount: number) {
        this._coinsTotal += coinCount
    }

    removeCoins(coinCount: number) {
        this._coinsTotal -= coinCount
    }

    get email() { return this._email; }

    get avatar() { return this._avatar }

    get title() { return this._title }

    get coins() { return this._coinsTotal }

    get kudos() { return this._kudosTotal }

    get lifetimeGames() { return this._gamesTotal }

    get lifetimeKudos() { return this._kudosOverall }

    get unlockedTitleFirstWords() { return this._unlockedTitleFirstWords }

    get unlockedTitleSecondWords() { return this._unlockedTitleSecondWords }

    get unlockedAvatars() { return this._unlockedAvatars }

    get strikes() { return this._strikes }

    set email(newEmail: string) {
        if (newEmail != null && newEmail != '')
            this._email = newEmail
    }

    set avatar(newAvatar: string) {
        if (newAvatar != null && newAvatar != '')
            this._avatar = newAvatar
    }

    set title(newTitle: string) {
        if (newTitle != null && newTitle != '')
            this._title = newTitle
    }
}

File: Politio.Api/Politio.Api/Data/Player.cs
using Microsoft.IdentityModel.Tokens;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Politio.Api.Data;

public class MyString
{
    [Key]
    public int Id { get; set; }
    public string? Value { get; set; }

    public MyString() { }
    public MyString(int id, string value)
    {
        Id = id;
        Value = value;
    }
}

public class Player
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }
    [JsonIgnore]public string Email { get; set; }
    [JsonIgnore]public string Password { get; set; }
    public string Title { get; set; }
    public string Avatar { get; set; }
    public int CoinsTotal { get; set; }
    public int GudosTotal { get; set; }
    public int GamesTotal { get; set; }
    public int KudosOverall { get; set; }
    public int Authoritarian { get; set; }
    public int Left { get; set; }
    public int Libertarian { get; set; }
    public int Right { get; set; }
    public List<MyString> UnlockedTitleFirstWords { get; }
    public List<MyString> UnlockedTitleSecondWords { get; }
    public List<MyString> UnlockedAvatars { get; }
    public int Strikes { get; set; }
    public string Theme { get; set; }
    public bool Activated { get; set; }
    public string ActualStanding { get; set; }

    public Player(string email, string password)
    {
        Email = email;
        Password = password;
        Title = "Scared Banana";
        Avatar = "starting_avatar_1.png";
        CoinsTotal = 0;
        GudosTotal = 0;
        GamesTotal = 0;
        KudosOverall = 0;
        Authoritarian = 0;
        Left = 0;
        Libertarian = 0;
        Right = 0;
        UnlockedTitleFirstWords = new List<MyString> { new MyString(1, "Angry"), new MyString(2, "Helpful"), new MyString(3, "Slimy"),
                                                        new MyString(4, "Scared") };
        UnlockedTitleSecondWords = new List<MyString> { new MyString(1, "Banana"), new MyString(2, "Explorer"), new MyString(3, "Marshmallow"), 
                                                         new MyString(4, "Racer") }; ;
        UnlockedAvatars = new List<MyString> { new MyString(1, "starting_avatar_1.png"), new MyString(2, "starting_avatar_2.png"), 
                                                new MyString(3, "starting_avatar_3.png"), new MyString(4, "starting_avatar_4.png") };
        Strikes = 0;
        Theme = "light";
        Activated = false;
        ActualStanding = "Unknown";
    }

    public string GetData()
    {
        return JsonSerializer.Serialize(this);
    }

    public bool CheckPassword(string password)
    {
        return Password.CompareTo(password) == 0;
    }

    public void IncAuthoritarian() { Authoritarian++; }
    public void IncLeft() { Left++; }
    public void IncLibertarian() { Libertarian++; }
    public void IncRight() { Right++; }

    public void AddTitleFirstWords(string[] newWords)
    {
        if (!newWords.IsNullOrEmpty())
            foreach (string word in newWords)
                UnlockedTitleFirstWords.Add(new MyString(UnlockedTitleFirstWords.Count + 1, word));

        throw new ArgumentNullException(nameof(newWords));
    }
    public void AddTitleSecondWords(string[] newWords)
    {
        if (!newWords.IsNullOrEmpty())
            foreach (string word in newWords)
                UnlockedTitleSecondWords.Add(new MyString(UnlockedTitleSecondWords.Count + 1, word));

        throw new ArgumentNullException(nameof(newWords));
    }

    public void AddAvatar(string avatar)
    {
        if (avatar is not null)
            UnlockedAvatars.Add(new MyString(UnlockedAvatars.Count + 1, avatar));
        
        throw new ArgumentNullException(nameof(avatar));
    }
}
